### DISCUSSION TITLE
understanding compilation errors? I don't....

### DISCUSSION THREAD

**Discussion by ClaireCJS** [microsoft/cascadia-code#777] [General]
Long story short — There's a ligature I don't like, so I forked the whole font and tried to rebuild it without that ligature.  

However, the resulting font was unusable because it was called CascadiaCode and I already have a CascadiaCode — yours. So I renamed every value and every filename from CascadiaCode to CascadiaCodeClaire 

Suddenly, I can no longer compile it. I'm sure it's a dumb error I made, but the error message just... isn't helpful... so I can't grok it to figure out how to fix it.

From what I can tell, if I can't figure out this error, I throw away about 90 minutes' work and poking around trying to remove the "***" ligature and failing.

I really don't want to fail. I want to get through this.

So this is the error I get ... recall it built just fine before I changed the name, and the name was the only thing I changed to start getting this error....

```
build.py
[Cascadia Code Claire Regular] Compiling
[Cascadia Mono Claire Regular] Compiling
[Cascadia Code Claire Italic] Compiling
[Cascadia Mono Claire Italic] Compiling
[Cascadia Mono Claire PL Italic ExtraLight Italic] Merging PL glyphs
[Cascadia Code Claire PL Italic ExtraLight Italic] Merging PL glyphs
[Cascadia Code Claire NF Italic ExtraLight Italic] Merging PL glyphs
[Cascadia Mono Claire NF Italic ExtraLight Italic] Merging PL glyphs
[Cascadia Code Claire NF ExtraLight] Merging PL glyphs
[Cascadia Mono Claire PL ExtraLight] Merging PL glyphs
[Cascadia Code Claire PL ExtraLight] Merging PL glyphs
[Cascadia Mono Claire NF ExtraLight] Merging PL glyphs
[Cascadia Code Claire NF Italic ExtraLight Italic] Merging NF glyphs
[Cascadia Mono Claire NF Italic ExtraLight Italic] Merging NF glyphs
[Cascadia Code Claire PL Italic Italic] Merging PL glyphs
[Cascadia Mono Claire PL Italic Italic] Merging PL glyphs
[Cascadia Mono Claire NF ExtraLight] Merging NF glyphs
[Cascadia Mono Claire PL Regular] Merging PL glyphs
[Cascadia Code Claire NF ExtraLight] Merging NF glyphs
[Cascadia Code Claire PL Regular] Merging PL glyphs
[Cascadia Mono Claire PL Italic Bold Italic] Merging PL glyphs
[Cascadia Code Claire PL Italic Bold Italic] Merging PL glyphs
[Cascadia Code Claire PL Bold] Merging PL glyphs
[Cascadia Mono Claire PL Bold] Merging PL glyphs
[Cascadia Mono Claire PL Italic] Compiling
[Cascadia Code Claire PL Italic] Compiling
Exception in thread Thread-3 (_handle_results):
Traceback (most recent call last):
  File "C:\ProgramData\anaconda3\lib\threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "C:\ProgramData\anaconda3\lib\threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "C:\ProgramData\anaconda3\lib\multiprocessing\pool.py", line 579, in _handle_results
    task = get()
  File "C:\ProgramData\anaconda3\lib\multiprocessing\connection.py", line 251, in recv
    return _ForkingPickler.loads(buf.getbuffer())
TypeError: FeatureLibError.__init__() missing 1 required positional argument: 'location'
[Cascadia Mono Claire PL Regular] Compiling
[Cascadia Code Claire PL Regular] Compiling
```

I definitely need help... Or better error messages.

**Comment by aaronbell** [comment]
Hi. The error message that you’re getting is a fontTools related one and while I am sure it appears incomprehensible, it is actually quite clear. Essentially it appears you have removed the ligature glyph from the font, but have not removed the reference to the ligature glyph from the OT feature code. So when the code is compiled at build time, it can’t find the glyph and issues a fail.

So you need to comment out, or remove this code from the files titled “calt” in the `sources/features` folder (you can ignore the one with “mono” in the name):
```
lookup asterisk_asterisk_asterisk {
  sub asterisk_asterisk_asterisk.liga LIG  asterisk'  by LIG;
  ignore sub asterisk_asterisk.liga' LIG asterisk slash;
  sub asterisk_asterisk.liga'         LIG  asterisk   by asterisk_asterisk_asterisk.liga;
} asterisk_asterisk_asterisk;
```

And actually if you wanted to avoid having to mess with Visual TrueType (VTT), you could not remove the glyph and *only* remove the above code. That will get rid of the ligature (no substitution code, no ligature replacement) without affecting glyph order and thus needing to re-run VTT compile.

**Reply by ClaireCJS** [reply]
> Hi. The error message that you’re getting is a fontTools related one and while I am sure it appears incomprehensible, it is actually quite clear. Essentially it appears you have removed the ligature glyph from the font

I guess — what do you mean by that?  The asterisk_asterisk_asterisk.liga.glif files are still all present.


> 
> So you need to comment out, or remove this code from the files titled “calt” in the `sources/features` folder (you can ignore the one with “mono” in the name):
> 
> ```
> lookup asterisk_asterisk_asterisk {
>   sub asterisk_asterisk_asterisk.liga LIG  asterisk'  by LIG;
>   ignore sub asterisk_asterisk.liga' LIG asterisk slash;
>   sub asterisk_asterisk.liga'         LIG  asterisk   by asterisk_asterisk_asterisk.liga;
> } asterisk_asterisk_asterisk;
> ```

I had done that, but I missed doing it in italics.

> 
> And actually if you wanted to avoid having to mess with Visual TrueType (VTT), you could not remove the glyph and _only_ remove the above code. That will get rid of the ligature (no substitution code, no ligature replacement) without affecting glyph order and thus needing to re-run VTT compile.

That's a great tip! I don't want to recompile anything, I just want to remove.

However, the error message now appears to me to be even more incomprehensible than I originally thought, because I'm still getting it despite the cause you said not being present. 

I think a proper error message would say what file is being read when it's missing, which would circumvent this whole conversation being necessary. I think it's bad programming practice myself. The fact that one has to guess as to what it means and it doesn't lead one precisely to the error? Bad.

So yeah. I did the things here, and I have the exact same error.

I mucked about with a lot of files, but those glif files are not missing, so them missing is not the cause of the error.

The error still persists.

I suppose I can try nuking the entire thing & starting over, doing ONLY that, and all the name changes (which were annoying), just to make sure nothing else got mucked about... But the fact that it wasn't what you said speaks to me about the quality of the error message.

Someone should write a simple guide on how to remove ligatures.  It's not as easy as one thinks. For instance, you can't just clone this repo and recompile, or you'll have a name collision. And I'm not sure just where the name needs to be changed, other than "everywhere", which takes like 20 minutes....

**Reply by aaronbell** [reply]
> I guess — what do you mean by that? The asterisk_asterisk_asterisk.liga.glif files are still all present.

Oh, it sounded like you'd deleted the file entirely? Maybe I don't totally understand the actual changes that you made to the files. 

> The fact that one has to guess as to what it means and it doesn't lead one precisely to the error? Bad.

I can understand your frustrations. I recall that there has been complaints that the toolchain does not produce easy to understand errors for those who haven't worked with it, or much with fonts before. The error indicates that the problem is related to the feature code, but perhaps if you could give me more information about what changes you made, or can point me to the code that you've modified, I can take a look? I remembered that there are another couple places where the `asterisk_asterisk_asterisk.liga` glyph is referenced in the feature code—specifically `header.fea` and `header_italic.fea` but if you haven't removed the `.glif` file for the glyph than those shouldn't be an issue. 

> For instance, you can't just clone this repo and recompile, or you'll have a name collision. 

Have you considered using the PL or NF variants instead? I believe Windows Terminal only uses the base `Cascadia Mono` and `Cascadia Code` so switching to one of the other variants would probably solve your problem. 

> And I'm not sure just where the name needs to be changed, other than "everywhere", which takes like 20 minutes....

Yes, the build script is not great for this. I've got a new version of the font and build script that should make it more straightforward to set the font name in the future.
